### PR TITLE
Remove package:test dependency from flutter_test

### PR DESCRIPTION
**Pull request by jonahwilliams**
Note: flutter still has all of the transitive dependencies of test due to mockito, I am filling a follow up PR now that test and test_api are on pub.

EDIT: mockito has landed (per discussion today). To avoid the chances of breaking users we should land this sooner rather than later

Both changes will have to wait until everything is sync'd internally as well.

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 1]
```diff
@@ -7,16 +7,30 @@ import 'dart:convert';
...
-
-import 'package:test/src/backend/runtime.dart'; // ignore: implementation_imports
-import 'package:test/src/backend/suite_platform.dart'; // ignore: implementation_imports
-import 'package:test/src/runner/plugin/platform.dart'; // ignore: implementation_imports
-import 'package:test/src/runner/plugin/hack_register_platform.dart' as hack; // ignore: implementation_imports
+import 'package:stack_trace/stack_trace.dart';
+
+import 'package:test_api/src/backend/runtime.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/suite_platform.dart'; // ignore: implementation_imports
+import 'package:test_api/src/runner/platform.dart'; // ignore: implementation_imports
```
Comment (grouma) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: I don't think having these "runner" files within `test_api` is providing any value and is in fact quite confusing. My suggestion is that we create a a new `test_core` package which contains the very basic running functionality. `flutter_tools` can depend only on `test_core` which will reduce the overall transitive dependencies. `flutter_test` can continue depending only on `test_api`. This is how it's structured within Google3.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: That sounds like a good approach, we'll talk tomorrow

**Comment by jonahwilliams** [comment]
Discussed offline with @grouma, he is going to land some changes to pull the runner sources out of test_api and into a new test_core package. This won't effect the dependencies for flutter_test, but will allow us to remove the package:test dependency of flutter_tool. Landing this internally may take some time.

**Comment by jonahwilliams** [comment]
Mockito has been updated. Still pending on internal sync and test_core

File: packages/flutter_test/lib/src/matchers.dart [hunk 2]
```diff
@@ -9,9 +9,9 @@ import 'dart:ui' as ui;
 import 'dart:ui';
 
 import 'package:meta/meta.dart';
-import 'package:test/test.dart' hide TypeMatcher, isInstanceOf;
-import 'package:test/test.dart' as test_package show TypeMatcher;
-import 'package:test/src/frontend/async_matcher.dart'; // ignore: implementation_imports
+import 'package:test_api/test_api.dart' hide TypeMatcher, isInstanceOf;
+import 'package:test_api/test_api.dart' as test_package show TypeMatcher;
+import 'package:test_api/src/frontend/async_matcher.dart'; // ignore: implementation_imports
```
Comment (dnfield) [inline packages/flutter_test/lib/src/matchers.dart]: Any reason for us to not just export this from test_api?

Comment (jonahwilliams) [reply packages/flutter_test/lib/src/matchers.dart]: This was done previously to hide a breaking change in the package:test matchers.

**Comment by dnfield** [comment]
From what I can tell LGTM - not 100% sure about what some of the additions are doing though.

**Comment by jonahwilliams** [comment]
There is another PR needed to merge in https://pub.dartlang.org/packages/test_core . This will require another mockito bump so I think we should do it separately

**Review by tvolkert — COMMENTED** [review]
LGTM, but deferring to @grouma on the extraction of the runner logic

File: packages/flutter_test/lib/src/runner.dart [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+  });
+}
+
+/// Returns a serialized test suite.
+StreamChannel<dynamic> serializeSuite(Function getMain(), {bool hidePrints = true, Future<dynamic> beforeLoad()}) {
+  return RemoteListener.start(getMain, hidePrints: hidePrints, beforeLoad: beforeLoad);
+}
+
+/// This is a listener.
+class RemoteListener {
```
Comment (grouma) [inline packages/flutter_test/lib/src/runner.dart]: This logic is really specific to the internals of the runner. Sorry I didn't catch this the first time around but I think we can do a large clean up here if we do the following:

test_core/lib/src/runner/remote_listener.dart -> test_api/lib/src/remote_listener.dart
test_core/lib/src/runner/suite_channel_manager.dart -> test_api/lib/src/suite_channel_manager.dart

This way you won't have to fork this file and flutter_test still only depends on test_api. Thoughts?

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 4]
```diff
@@ -4,19 +4,37 @@
...
+import 'package:test_api/src/runner/hack_register_platform.dart' as hack; // ignore: implementation_imports
+import 'package:test_api/src/runner/runner_suite.dart'; // ignore: implementation_imports
+import 'package:test_api/src/runner/suite.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/group.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/test.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/live_test.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/metadata.dart'; // ignore: implementation_imports
+import 'package:test_api/src/runner/environment.dart'; // ignore: implementation_imports
```
Comment (grouma) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: test_core is now available. Up to you if you want to update this PR or do a follow up to get onto the new package.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 5]
```diff
@@ -135,14 +153,9 @@ String generateTestBootstrap({
 import 'dart:convert';
 import 'dart:io';  // ignore: dart_io_import
 
-// We import this library first in order to trigger an import error for
-// package:test (rather than package:stream_channel) when the developer forgets
-// to add a dependency on package:test.
-import 'package:test/src/runner/plugin/remote_platform_helpers.dart';
```
Comment (grouma) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Note with my proposed changes this file basically stays the same. We will now depend on the sources from test_api and test_core though.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Perfect, I'll roll that in with the latest test_api/test_core

**Comment by jonahwilliams** [comment]
Removed forked sources, just waiting on mockito to publish 4.0

**Review by grouma — APPROVED** [review]

**Comment by jonahwilliams** [comment]
Some of the deps are still wrong until I update mockito

**Comment by jonahwilliams** [comment]
I've rolled in and pinned mockito so that users of flutter_test cannot pull in older versions that will break. This is still a reduction in the overall number of pinned deps

**Comment by jonahwilliams** [comment]
Well, I accidentally merged an unrelated branch X.X

**Comment by dnfield** [comment]
This is breaking the devicelab.

**Comment by jonahwilliams** [comment]
looking at it currently

**Pull request merged**